src/Transaction.ts

```diff
@@ -0,0 +1,97 @@
+/*
+   Copyright 2017-2023 Norman Breau
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import { IDatabaseConnection } from './IDatabaseConnection';
+import {IQueryable} from './IQueryable';
+import { Application } from './Application';
+import { IsolationLevel } from './IsolationLevel';
+import { InternalError } from './InternalError';
+import { DeadLockError } from './DeadLockError';
+import { InvalidValueError } from './InvalidValueError';
+
+const TAG: string = 'Transaction';
+
+export type ITransactionExecutor = (connection: IDatabaseConnection) => Promise<void>;
+
+/**
+ * A class encapsulating an entire transaction from beginning to commitment.
+ * 
+ * This encapsulates a series of Query steps to conduct for the transaction.
+ * Should the transaction fail due to a deadlock, the transaction will automatically
+ * 
+ */
+export class Transaction implements IQueryable<void> {
+    private $retryLimit: number;
+    private $application: Application;
+    private $isolationLevel: IsolationLevel;
+    private $executor: ITransactionExecutor;
+
+    public constructor(app: Application, executor: ITransactionExecutor, retryLimit: number = Infinity, isolationLevel: IsolationLevel = IsolationLevel.REPEATABLE_READ) {
+        this.$application = app;
+        this.$executor = executor;
+
+        if (retryLimit === null || retryLimit === undefined) {
+            retryLimit = Infinity;
```

Comment:
I wonder if retryLimit should be opt in instead of opt out.

Currently, the default behaviour, is to never let a transaction effectively error out. This in itself would be a breaking change.

Or maybe the retry limit default should be something that is realistically hittable. Like 10 or something.

Reply:
Oh wait, retry limit only applies to deadlock errors. I wonder if this is a good idea.

-Will the API be confusing? (retry might suggest on any error.)
-Are there other errors that would be a good idea to retry on?

Reply:
> Currently, the default behaviour, is to never let a transaction effectively error out. This in itself would be a breaking change.

This is being introduced in a major version bump anyway.

> -Are there other errors that would be a good idea to retry on?

AFAIK, deadlocks is the only error that is safe to retry. Any other error indicates a problem, either a developer error, or a concurrency error where retrying will not change the result (e.g. an unique constraint failure)

Reply:
> Will the API be confusing? (retry might suggest on any error.)

Also explicitly stated in the code docs that it's for handling deadlocks.